source/coap_message_handler.c

```diff
@@ -209,6 +210,10 @@ int16_t coap_message_handler_coap_msg_process(coap_msg_handler_t *handle, int8_t
             transaction_ptr->service_id = coap_service_id_find_by_socket(socket_id);
             transaction_ptr->msg_id = coap_message->msg_id;
             transaction_ptr->client_request = false;// this is server transaction
+            if (dst_addr_ptr) {
+                memcpy(transaction_ptr->local_address, dst_addr_ptr, 16);
+                transaction_ptr->have_local_address = true;
+            }
```

Comment:
This is the only place you need a condition - `else memcpy(transaction_ptr->local_address, ns_in6addr_any, 16)`.

Actually, this code isn't valid, you are passing `dst_addr_ptr`  as [static 16] which means it can't be null on entry. Compiler could be optimising this test out.

Either require that they pass "any" or turn it into a normal pointer parameter.